File: JavaScript/examples/base64_utils_example.js
/**
 * Base64 Utilities - 使用示例
 * 
 * 本文件演示如何使用 base64_utils 模块的各种功能。
 * 
 * 运行方式:
 * - Node.js: node base64_utils_example.js
 * - 浏览器: 在 HTML 中引入 <script src="../base64_utils/mod.js"></script>
 *           然后在控制台运行示例代码
 */

// 加载模块（Node.js 环境）
const Base64Utils = (typeof require !== 'undefined') 
  ? require('../base64_utils/mod.js') 
  : window;

const { encode, decode, toUrlSafe, fromUrlSafe, isValid, toUint8Array, fromUint8Array, encodedLength } = Base64Utils;

console.log('========================================');
console.log('Base64 Utilities - Usage Examples');
console.log('========================================\n');

// ========================================
// 1. 基本编码和解码
// ========================================
console.log('--- 1. Basic Encoding and Decoding ---\n');

const originalText = 'Hello, World! 你好，世界！';
console.log('Original:', originalText);

// 编码为 Base64
const encoded = encode(originalText);
console.log('Encoded:', encoded);

// 解码回原始文本
const decoded = decode(encoded);
console.log('Decoded:', decoded);
console.log('Match:', originalText === decoded ? '✓' : '✗');

// ========================================
// 2. URL 安全 Base64
// ========================================
console.log('\n--- 2. URL-Safe Base64 ---\n');

// 标准 Base64 包含 + 和 /，在 URL 中需要编码
const standardBase64 = encode('>>>???');
console.log('Standard Base64:', standardBase64);

// 转换为 URL 安全格式（+ -> -, / -> _, 移除 =）
const urlSafe = toUrlSafe(standardBase64);
console.log('URL-Safe:', urlSafe);

// 从 URL 安全格式转回标准格式
const backToStandard = fromUrlSafe(urlSafe);
console.log('Back to Standard:', backToStandard);

// 直接编码为 URL 安全格式
const directUrlSafe = encode('Hello World', { urlSafe: true, pad: false });
console.log('Direct URL-Safe:', directUrlSafe);

// ========================================
// 3. 验证 Base64 字符串
// ========================================
console.log('\n--- 3. Validation ---\n');

const validString = 'SGVsbG8gV29ybGQ=';
const invalidString = 'Invalid!!!';

console.log(`isValid('${validString}'):`, isValid(validString) ? '✓ Valid' : '✗ Invalid');
console.log(`isValid('${invalidString}'):`, isValid(invalidString) ? '✓ Valid' : '✗ Invalid');

// 验证 URL 安全格式
const urlSafeString = 'SGVsbG8gV29ybGQ';
console.log(`isValid('${urlSafeString}', { urlSafe: true }):`, 
  isValid(urlSafeString, { urlSafe: true }) ? '✓ Valid' : '✗ Invalid');

// ========================================
// 4. 处理二进制数据
// ========================================
console.log('\n--- 4. Binary Data Handling ---\n');

// 将 Base64 转换为 Uint8Array
const base64Data = 'SGVsbG8gV29ybGQ=';
const uint8Array = toUint8Array(base64Data);
console.log('Base64 to Uint8Array:', base64Data);
console.log('Uint8Array:', Array.from(uint8Array));

// 将 Uint8Array 编码回 Base64
const backToBase64 = fromUint8Array(uint8Array);
console.log('Back to Base64:', backToBase64);

// 创建自定义二进制数据
const customBytes = new Uint8Array([0x48, 0x65, 0x6C, 0x6C, 0x6F]); // "Hello"
const customBase64 = fromUint8Array(customBytes);
console.log('Custom bytes to Base64:', customBase64);

// ========================================
// 5. 计算编码后长度
// ========================================
console.log('\n--- 5. Calculate Encoded Length ---\n');

const testStrings = ['A', 'AB', 'ABC', 'Hello', 'Hello World'];
for (const str of testStrings) {
  const withPad = encodedLength(str);
  const withoutPad = encodedLength(str, { pad: false });
  console.log(`'${str}' -> With padding: ${withPad}, Without padding: ${withoutPad}`);
}

// ========================================
// 6. 实际应用场景
// ========================================
console.log('\n--- 6. Practical Use Cases ---\n');

// 场景 1: 在 URL 中传递数据
console.log('Use Case 1: Passing data in URL');
const userData = { name: 'John', age: 30 };
const jsonString = JSON.stringify(userData);
const urlSafeData = encode(jsonString, { urlSafe: true, pad: false });
const url = `https://example.com/data?payload=${urlSafeData}`;
console.log('URL:', url.substring(0, 70) + '...');

// 场景 2: 处理 Unicode 字符（如 Emoji）
console.log('\nUse Case 2: Encoding Unicode/Emoji');
const emoji = '🎉🎊🎁';
const emojiEncoded = encode(emoji);
const emojiDecoded = decode(emojiEncoded);
console.log('Original:', emoji);
console.log('Encoded:', emojiEncoded);
console.log('Decoded:', emojiDecoded);

// 场景 3: 处理多语言文本
console.log('\nUse Case 3: Multi-language Text');
const multiLang = 'Hello 你好 こんにちは 안녕하세요';
const multiEncoded = encode(multiLang);
const multiDecoded = decode(multiEncoded);
console.log('Original:', multiLang);
console.log('Encoded:', multiEncoded);
console.log('Decoded:', multiDecoded);
console.log('Match:', multiLang === multiDecoded ? '✓' : '✗');

// ========================================
// 7. 无填充编码
// ========================================
console.log('\n--- 7. Encoding Without Padding ---\n');

const text = 'Hello';
const withPadding = encode(text);
const withoutPadding = encode(text, { pad: false });

console.log('Original:', text);
console.log('With padding:', withPadding);
console.log('Without padding:', withoutPadding);

// 解码时两种格式都可以正确解码
console.log('Decode with padding:', decode(withPadding));
console.log('Decode without padding:', decode(withoutPadding));

// ========================================
// 8. 错误处理示例
// ========================================
console.log('\n--- 8. Error Handling ---\n');

// 尝试解码无效的 Base64
try {
  decode('This is not valid!!!');
} catch (e) {
  console.log('Caught error for invalid Base64:', e.message);
}

// 尝试对非字符串进行编码
try {
  encode(12345);
} catch (e) {
  console.log('Caught error for non-string input:', e.message);
}

// 使用 isValid 进行安全验证
const userInput = 'UserInput123';
if (isValid(userInput)) {
  console.log('Valid Base64, decoding:', decode(userInput));
} else {
  console.log('Invalid Base64, skipping decode');
}

console.log('\n========================================');
console.log('Examples completed!');
console.log('========================================');

File: C++/examples/base64_utils_example.cpp
/**
 * @file base64_utils_example.cpp
 * @brief Usage examples for Base64 Utilities
 *
 * Demonstrates various use cases for Base64 encoding/decoding.
 *
 * @author AllToolkit Contributors
 * @version 1.0.0
 * @license MIT
 */

#include <iostream>
#include <vector>
#include <cstdint>
#include "../base64_utils/mod.hpp"

using namespace alltoolkit;

int main() {
    std::cout << "==============================================" << std::endl;
    std::cout << "Base64 Utilities - Usage Examples" << std::endl;
    std::cout << "==============================================" << std::endl;

    // =============================================================================
    // Example 1: Basic String Encoding/Decoding
    // =============================================================================
    std::cout << "\n1. Basic String Encoding/Decoding" << std::endl;
    std::cout << "-----------------------------------" << std::endl;

    std::string text = "Hello, World!";
    std::string encoded = Base64Utils::encode(text);
    std::string decoded = Base64Utils::decode(encoded);

    std::cout << "Original: " << text << std::endl;
    std::cout << "Encoded:  " << encoded << std::endl;
    std::cout << "Decoded:  " << decoded << std::endl;

    // =============================================================================
    // Example 2: Binary Data Encoding
    // =============================================================================
    std::cout << "\n2. Binary Data Encoding" << std::endl;
    std::cout << "------------------------" << std::endl;

    std::vector<uint8_t> binaryData = {0x00, 0x01, 0x02, 0x03, 0xFF, 0xFE};
    std::string binaryEncoded = Base64Utils::encode(binaryData);
    std::vector<uint8_t> binaryDecoded = Base64Utils::decodeToBytes(binaryEncoded);

    std::cout << "Original bytes: ";
    for (auto b : binaryData) {
        std::cout << std::hex << "0x" << static_cast<int>(b) << " ";
    }
    std::cout << std::dec << std::endl;
    std::cout << "Encoded: " << binaryEncoded << std::endl;
    std::cout << "Decoded bytes: ";
    for (auto b : binaryDecoded) {
        std::cout << std::hex << "0x" << static_cast<int>(b) << " ";
    }
    std::cout << std::dec << std::endl;

    // =============================================================================
    // Example 3: URL-safe Base64
    // =============================================================================
    std::cout << "\n3. URL-safe Base64" << std::endl;
    std::cout << "-------------------" << std::endl;

    std::string special = "user+name@example.com";
    std::string standardB64 = Base64Utils::encode(special);
    std::string urlSafeB64 = Base64Utils::encode(special, true, false);

    std::cout << "Original: " << special << std::endl;
    std::cout << "Standard Base64: " << standardB64 << std::endl;
    std::cout << "URL-safe Base64: " << urlSafeB64 << std::endl;

    // Demonstrate conversion
    std::string converted = Base64Utils::toUrlSafe(standardB64, false);
    std::cout << "Converted to URL-safe: " << converted << std::endl;

    // =============================================================================
    // Example 4: Padding Control
    // =============================================================================
    std::cout << "\n4. Padding Control" << std::endl;
    std::cout << "-------------------" << std::endl;

    std::string shortText = "Hi";
    std::string withPadding = Base64Utils::encode(shortText, false, true);
    std::string noPadding = Base64Utils::encode(shortText, false, false);

    std::cout << "Original: " << shortText << std::endl;
    std::cout << "With padding: " << withPadding << std::endl;
    std::cout << "No padding:   " << noPadding << std::endl;

    // Both decode to the same value
    std::cout << "Decoded (with padding): " << Base64Utils::decode(withPadding) << std::endl;
    std::cout << "Decoded (no padding):   " << Base64Utils::decode(noPadding) << std::endl;

    // =============================================================================
    // Example 5: Validation
    // =============================================================================
    std::cout << "\n5. Validation" << std::endl;
    std::cout << "--------------" << std::endl;

    std::vector<std::string> testStrings = {
        "SGVsbG8=",      // Valid
        "Zm9vYmFy",      // Valid
        "Invalid!",      // Invalid
        "Z",             // Invalid (wrong length)
        "SGVsbG8-",      // URL-safe valid
        "Zm9v_"          // URL-safe valid
    };

    for (const auto& str : testStrings) {
        bool standard = Base64Utils::isValid(str, false);
        bool urlSafe = Base64Utils::isValid(str, true);
        std::cout << "\"" << str << "\" - Standard: " << (standard ? "Valid" : "Invalid")
                  << ", URL-safe: " << (urlSafe ? "Valid" : "Invalid") << std::endl;
    }

    // =============================================================================
    // Example 6: Length Calculations
    // =============================================================================
    std::cout << "\n6. Length Calculations" << std::endl;
    std::cout << "-----------------------" << std::endl;

    std::vector<size_t> sizes = {1, 2, 3, 10, 100, 1000};
    for (size_t size : sizes) {
        size_t encodedLen = Base64Utils::encodedLength(size, true);
        size_t noPadLen = Base64Utils::encodedLength(size, false);
        std::cout << "Input " << size << " bytes -> Encoded: " << encodedLen
                  << " (no pad: " << noPadLen << ")" << std::endl;
    }

    // =============================================================================
    // Example 7: Practical Use Case - Simple Token Generation
    // =============================================================================
    std::cout << "\n7. Practical: Simple Token Generation" << std::endl;
    std::cout << "--------------------------------------" << std::endl;

    // Simulate a simple token: user_id:timestamp
    std::string userId = "user123";
    std::string timestamp = "1704067200";
    std::string tokenData = userId + ":" + timestamp;
    std::string token = Base64Utils::encode(tokenData, true, false);

    std::cout << "User ID: " << userId << std::endl;
    std::cout << "Timestamp: " << timestamp << std::endl;
    std::cout << "Token: " << token << std::endl;

    // Decode and verify
    std::string decodedToken = Base64Utils::decode(token, true);
    std::cout << "Decoded: " << decodedToken << std::endl;

    // =============================================================================
    // Example 8: Error Handling
    // =============================================================================
    std::cout << "\n8. Error Handling" << std::endl;
    std::cout << "------------------" << std::endl;

    try {
        Base64Utils::decode("Invalid!Base64@");
    } catch (const std::invalid_argument& e) {
        std::cout << "Caught expected error: " << e.what() << std::endl;
    }

    // =============================================================================
    // Example 9: Working with Raw Pointers
    // =============================================================================
    std::cout << "\n9. Raw Pointer Interface" << std::endl;
    std::cout << "-------------------------" << std::endl;

    const uint8_t rawData[] = {0x48, 0x65, 0x6C, 0x6C, 0x6F}; // "Hello"
    std::string rawEncoded = Base64Utils::encode(rawData, sizeof(rawData));
    std::cout << "Raw data encoded: " << rawEncoded << std::endl;

    // =============================================================================
    // Example 10: Round-trip Verification
    // =============================================================================
    std::cout << "\n10. Round-trip Verification" << std::endl;
    std::cout << "----------------------------" << std::endl;

    std::vector<std::string> testData = {
        "",
        "A",
        "AB",
        "ABC",
        "The quick brown fox jumps over the lazy dog",
        "1234567890!@#$%^&*()"
    };

    bool allPassed = true;
    for (const auto& data : testData) {
        std::string enc = Base64Utils::encode(data);
        std::string dec = Base64Utils::decode(enc);
        bool passed = (data == dec);
        allPassed = allPassed && passed;
        std::cout << "\"" << data.substr(0, 20) << (data.length() > 20 ? "..." : "")
                  << "\" -> " << (passed ? "PASS" : "FAIL") << std::endl;
    }

    std::cout << "\nAll tests: " << (allPassed ? "PASSED" : "FAILED") << std::endl;

    std::cout << "\n==============================================" << std::endl;
    std::cout << "Examples completed!" << std::endl;
    std::cout << "==============================================" << std::endl;

    return 0;
}